# Laboratorio de regresión logística

|                |   |
:----------------|---|
| **Nombre**     | Valeria Estefanía Milke Loera  |
| **Fecha**      | 16-03-2026  |
| **Expediente** |  739228 | 

In machine learning, Support Vector Machines (SVM) are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. It is mostly used in classification problems. In this algorithm, each data item is plotted as a point in p-dimensional space (where p is the number of features), with the value of each feature being the value of a particular coordinate. Then, classification is performed by finding the hyper-plane that best differentiates the two classes (or more if we have a multi class problem):

$$ f(x) = w^T \varphi(x) + b $$

where $\varphi: X \rightarrow F $ is a function that makes each input point $x$ correspond to a point in F, where F is a Hilbert space.

In addition to performing linear classification, SVMs can efficiently perform a non-linear classification, implicitly mapping their inputs into high-dimensional feature spaces (more specifically using the kernel trick, like the RBF funcion). 

[1]

OLS utilizes the squared residuals to fit the parameters. Large residuals caused by outliers may worsen the accuracy significantly.

Support Vectors use piecewise linear functions to counter this, in which a hyperparameter  $\epsilon$ called the margin lets errors that are less or equal to it be 0, and error larger than it be $e - \epsilon$. 

The problem to solve is:

\begin{split}
        \min_{w, b, \xi, \xi^*} \mathcal{P}_\epsilon(w, b, \xi) &= \frac{1}{2} w^T w + c \sum_{k=1}^{N} \xi_k \\
        \text{s.t. } & y_k [w^T \varphi(x_k) - b] \geq 1- \xi_k,\ \ k = 1, ..., N \\
        & \xi_k \geq 0,\ \ k = 1, ..., N
\end{split}


The most important question that arises when using a SVM is how to choose the correct hyperplane. Consider the following scenarios:

### Scenario 1

In this scenario there are three hyperplanes called A, B, and C. Now, the problem is to identify the hyperplane which best differentiates the stars and the circles.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_21-2.png" alt="what image shows"></center>

In this case, hyperplane B separates the stars and the circle betters, hence it is the correct hyperplane.


### Scenario 2

Now take another scenario where all three hyperplanes are segregating classes well. The question that arises is how to choose the best hyperplane in this situation.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_4-2.png" alt="what image shows"></center>

In such scenarios, we calculate the margin (which is the distance between nearest data point and the hyperplane). The hyperplane with the largest margin will be considered as the correct hyperplane to classify the dataset.

Here C has the largest margin. Hence, it is considered as the best hyperplane.


### Kernels
Knowing 
$$ w = \sum_{k=1}^{N} \alpha_k y_k \varphi(x_k) $$

And
$$ y_{pred} = w^T \varphi(x) + b $$

Then 
$$ y_{pred} = (\sum_{k=1}^{N} \alpha_k y_k \varphi(x_k))^T \varphi(x) + b $$

Where $\varphi$ is a function that makes each input in $x$ correspond to a point in $F$ (a Hilbert space). This can be seen as processing and transforming the input featuers to keep the model's convexity. [2]

This also allows us to transform the inputs into another space where they might be more easily classified.

<center><img src=https://miro.medium.com/max/838/1*gXvhD4IomaC9Jb37tzDUVg.png alt="what image shows"></center>

## ROC and AUC

A ROC (Receiver Operating Characteristic) is a graph that shows how the classification model performs at the classification thresholds. 

ROC curves typically feature true positive rate on the Y axis, and false positive rate on the X axis. This means that the top left corner of the plot is the “ideal” point - a false positive rate of zero, and a true positive rate of one. This is not very realistic, but it does mean that a larger area under the curve (AUC) is usually better. [3]

True Positive Rate is a synonym for Recall and defined as:
$$ TPR = \frac{TP}{TP + FN} $$

False Positive Rate is a synonym for Specificity and defined as:

$$ FPR = \frac{FP}{FP + TN} $$

ROC curves are typically used in binary classification to study the output of a classifier. In order to extend ROC curve and ROC area to multi-label classification, it is necessary to binarize the output. One ROC curve can be drawn per label, but one can also draw a ROC curve by considering each element of the label indicator matrix as a binary prediction (micro-averaging).

E.g. If you lower a classification threshold, more items would be classified as positive, increasing False Positives and True Positives.

AUC stands for Area under the ROC.

## Ejercicio 1

- Utiliza el dataset `Iris`, modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').
- Modela con LogisticRegression.
- El método de Cross-Validation es K-Folds con $k=10$.
- Utiliza el AUC como métrico de Cross-Validation.
- Compara resultados.

In [1]:
from sklearn.datasets import load_iris
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import KFold
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_val_score
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder

In [2]:
iris = load_iris()
x = iris.data
y = iris.target

In [3]:
kernels = ["linear","poly","rbf","sigmoid"]

In [4]:
for k in kernels:
    model = SVC(kernel=k, probability=True)
    scores = cross_val_score(
        model,
        x,
        y,
        scoring="roc_auc_ovr",
        cv=10
    )
    print("Kernel:",k)
    print(scores)
    print("Promedio AUC:",scores.mean())

Kernel: linear
[1.         1.         1.         1.         1.         1.
 0.98666667 1.         1.         1.        ]
Promedio AUC: 0.9986666666666666
Kernel: poly
[1.         1.         1.         1.         1.         1.
 0.98666667 1.         1.         1.        ]
Promedio AUC: 0.9986666666666666
Kernel: rbf
[1.         1.         1.         1.         1.         0.98666667
 0.98666667 1.         1.         1.        ]
Promedio AUC: 0.9973333333333333
Kernel: sigmoid
[0.79333333 0.71333333 0.67333333 0.62666667 0.74666667 0.8
 0.65333333 0.76       0.72       0.76      ]
Promedio AUC: 0.7246666666666666


In [6]:
log_model = LogisticRegression(max_iter=1000)
scores_2 = cross_val_score(
    log_model,
    x,
    y,
    scoring='roc_auc_ovr',
    cv=10
)
print("logistic", scores_2)

logistic [1.         1.         1.         1.         1.         1.
 0.98666667 1.         1.         1.        ]


**Interpretación**

En el dataset Iris, los kernels linear, poly y rbf obtuvieron valores de AUC cercanos a 1, lo que indica una clasificación casi perfecta. Esto sucede porque las clases del dataset son fácilmente separables. El kernel sigmoid mostró un desempeño menor, con un AUC alrededor de 0.72. El modelo de Logistic Regression también obtuvo resultados muy altos, similares a los mejores kernels de SVM.

En conclusión, para el dataset Iris, tanto SVM con kernels linear, poly o rbf como Logistic Regression logran una clasificación casi perfecta, mientras que el kernel sigmoid presenta un desempeño inferior.

## Ejercicio 2
- Repite el ejercicio 1 con el dataset `Default`. Utiliza `default` como target.

In [7]:
df = pd.read_csv(r"C:\Users\valer\Default.csv")

In [8]:
df.head()

,default,student,balance,income
0,No,No,729.526495,44361.625074
1,No,Yes,817.180407,12106.134700
2,No,No,1073.549164,31767.138950
3,No,No,529.250605,35704.493940
4,No,No,785.655883,38463.495880


In [12]:
colnum = ["balance","income"]
colcat = ["student"]

x = df[colnum + colcat]
y = df["default"]

In [13]:
num_transformer = StandardScaler()
cat_transformer = OneHotEncoder()

In [14]:
preprocessor = ColumnTransformer(
    transformers=[
        ('num', num_transformer, colnum),
        ('cat', cat_transformer, colcat)
    ]
)

In [15]:
kernels = ["linear", "poly", "rbf","sigmoid"]

In [17]:
for i in kernels:
    model = SVC(probability=True, kernel=i)
    pipeline = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('model', model)
    ])
    scores = cross_val_score(
        pipeline,
        x,
        y,
        scoring='roc_auc',
        cv=10
    )
    print(i, scores)
    print("Promedio AUC:", scores.mean())
    print()

linear [0.94017737 0.85468961 0.96220739 0.93792109 0.93513209 0.93604086
 0.95080066 0.94452564 0.90588844 0.89836804]
Promedio AUC: 0.9265751191770837

poly [0.80906271 0.79583843 0.86070634 0.8245746  0.78969634 0.86431011
 0.8658143  0.82855316 0.83372914 0.79670564]
Promedio AUC: 0.826899076334737

rbf [0.80332801 0.7928614  0.86070634 0.87205039 0.80555294 0.87712701
 0.84419166 0.84131044 0.85692973 0.81320789]
Promedio AUC: 0.8367265800581855

sigmoid [0.70533672 0.60919432 0.82140955 0.76603679 0.74811194 0.76111686
 0.68302466 0.70509073 0.73502009 0.6623432 ]
Promedio AUC: 0.7196684854582827



**Interpretación**

Para este segundo ejercicio se repitió el procedimiento del ejercicio usando el dataset "Default", donde el objetivo es predecir si un cliente incumplirá el pago de su crédito. Para mejorar el desempeño y evitar errores, se aplicó un preprocesamiento de datos para las variables numéricas y categóricas y se escalaron los datos. 

Los resultados de los modelos SVM con diferentes kernels muestran que el linear tuvo el mejor desempeño (0.93), lo que sugiere que la relación entre las variables predictoras y la variable objetivo puede aproximarse adecuadamente mediante una frontera de decisión lineal. Por otro lado, los kernels rbf y poly también presentan un desempeño aceptable, aunque inferior al kernel linear, mientras que el kernel sigmoid muestra el peor desempeño entre los evaluados.

En conclusión, para el dataset Default, el modelo SVM con kernel linear es el que ofrece la mejor capacidad de clasificación según la métrica AUC.

# Addendum

Métricos disponibles para clasificación:
- ‘accuracy’
- ‘balanced_accuracy’
- ‘top_k_accuracy’
- ‘average_precision’
- ‘neg_brier_score’
- ‘f1’
- ‘f1_micro’
- ‘f1_macro’
- ‘f1_weighted’
- ‘f1_samples’
- ‘neg_log_loss’
- ‘precision’ etc.
- ‘recall’ etc.
- ‘jaccard’ etc.
- ‘roc_auc’
- ‘roc_auc_ovr’
- ‘roc_auc_ovo’
- ‘roc_auc_ovr_weighted’
- ‘roc_auc_ovo_weighted’
- ‘d2_log_loss_score’

# References

[1] Shigeo Abe.Support Vector Machines for Pattern Classification,2Ed.Springer-Verlag London,2010. ISBN978-1-84996-097-7. URLhttps://www.springer.com/gp/book/9781849960977.

[2] Johan A K Suykens, Tony Van Gestel, Jos De Brabanter, BartDe Moor, and Joos Vandewalle.Least Squares Support VectorMachines. World Scientific,2002. ISBN9789812381514. URLhttps://www.worldscientific.com/worldscibooks/10.1142/5089.

[3] Bradley, A. P. (1997). The use of the area under the ROC curve in the evaluation of machine learning algorithms. Pattern recognition, 30(7), 1145-1159. URL https://www.researchgate.net/post/how_can_I_interpret_the_ROC_curve_result